### ISSUE TITLE
No modular exponentiation?

### ISSUE THREAD

**Issue by piquan** [numpy/numpy#8804]
It would be useful if NumPy had a modular exponentiation function.  Besides performance improvements over `a**b%m`, it also allows computations where `a**b` would not fit in the datatype.

NumPy also seems to ignore Python's third argument to `pow` (the modulus), rather than raising an error if it's set.

    In [2]: pow(np.arange(10), 10, 3)
    Out[2]: 
    array([         0,          1,       1024,      59049,    1048576,
              9765625,   60466176,  282475249, 1073741824, 3486784401])

**Comment by seberg** [comment]
Yeah, the code even has a comment about it. It seems a bit funny that there is no obvious way to indicate that the modulo part is not supported, though. I mean, I am not sure you can ask python to do a manual modulo fallback at all.

**Comment by eric-wieser** [comment]
Would `if (modulo != Py_None) { Py_INCREF(Py_NotImplemented); return Py_NotImplemented; }` do the trick?

**Comment by seberg** [comment]
@eric-wieser dunno, might work, though it was not easy enough for me to figure out if it won't do odd stuff in other cases.

**Comment by eric-wieser** [comment]
I'll make a PR and see if tests pass. Modular power isn't expected to delegate to `__r*__` methods, so I think this is reasonably straightforward

**Comment by seberg** [comment]
Ah ok if it is not expected to delegate or so, then its probably fine.

**Comment by eric-wieser** [comment]
I've split this into a separate issue for the silent failure at #8805

**Comment by piquan** [comment]
If somebody comes across this entry and needs a workaround, here it is:

```
# Taken from Applied Cryptography via Wikipedia
# https://en.wikipedia.org/w/index.php?title=Modular_exponentiation&oldid=761445844#Right-to-left_binary_method
def modular_pow(base, exponent, modulus):
    if modulus == 1:
        return np.zeros_like(base)
    # FIXME Wikipedia says: Assert :: (modulus - 1) * (modulus - 1) does not overflow base
    # Not sure what it means by "base" here: the "base" argument, or the bit representation?
    # Schneier's original doesn't mention this assertion.
    result = np.ones_like(base)
    base = base % modulus
    while exponent > 0:
        if (exponent % 2 == 1):
            result = (result * base) % modulus
        exponent = exponent >> 1
        base = (base * base) % modulus
    return result
```

**Comment by eric-wieser** [comment]
@piquan: Numpy 1.13 will now at least error when the 3-argument version is used. (unless someone gets a PR in before release ;) )

**Comment by piquan** [comment]
Great, thanks!

Thanks for the scalar note; I realized the problem with the workaround after I posted it, but didn't feel it would be worth a followup post.  The research I'm doing now uses scalar exponents, so the idea of having vector exponents didn't even occur to me.  (Same with moduli, but at first glance it looks like it would work by coincidence.)  I don't know NumPy well enough to write efficient conditionalized vector code (where only some of the elements will take a branch), so I may not tackle this just yet.

I did take another look at the FIXME comment, and it is indeed to assert that (modulus-1)**2 doesn't overflow the representation.

**Comment by jeffrey-cochran** [comment]
> That workaround only works for scalar `exponent`

Never thought I would need this until I did hahaha. Is there a workaround for array exponents?

**Comment by lorepieri8** [comment]
Any plans to see this live?

**Comment by sh1boot** [comment]
> I did take another look at the FIXME comment, and it is indeed to assert that (modulus-1)**2 doesn't overflow the representation.

It's not quite right, btw.  The maximum result is actually: (modulus-1)**2 * base.size[0]

**Comment by piquan** [comment]
> It's not quite right, btw. The maximum result is actually: (modulus-1)**2 * base.size[0]

I'm afraid I don't quite understand.  `ndarray.size` is a Python integer, so I don't know what you mean by `base.size[0]`.  Could you clarify that?

I also don't know what's wrong with using `(modulus-1)**2` by itself.  The goal of this assertion is to ensure that the intermediate value of `base*base` (in the last line of the loop) won't overflow the dtype.  Since `base` is always less than `modulus`, then you can make that assertion by considering the value of `(modulus-1)**2`, I expect.

In this routine, each element is handled independently — there’s no operation that multiplies or sums across the entire array in a way that would boost intermediate values beyond `(modulus − 1)**2`.  In other words, we perform modular exponentiation on each element separately, so each element’s maximum intermediate result is always `(modulus - 1)**2`. That’s why the total size or shape of the array (like base.size) isn’t relevant for preventing overflow.

I didn't include this assertion in my sample workaround above.  If you want to do so, I think you can use something like this:

```
# Get the maximum modulus.  This is safe even if modulus is a scalar,
# but if you know it's a scalar or a Python integer, you can simplify
# this accordingly.  We make it a Python arbitrary-precision integer to
# ensure that computing (modulus_max-1)**2 below won't overflow.
modulus_max = int(np.max(modulus))
# Get the maximum value that can be represented in base's dtype.
limit = np.iinfo(base.dtype).max
# Make sure it fits.
if (modulus_max - 1)**2 > limit:
    raise ValueError("Modular exponentiation may overflow the base's dtype.")
```

**Comment by sh1boot** [comment]
Oh.  I only just realised this isn't actually a bug about the matrix exponentiation.

Would matrix follow this same path with an overloaded multiplication operation?

**Comment by piquan** [comment]
Ah, yes, this is about the `pow` function, which is an element-wise umath operation.  Importantly, the `pow` function in Python accepts a modulus argument, which NumPy silently ignored prior to 1.13 (although [the docs](https://numpy.org/doc/2.2/reference/arrays.ndarray.html#arithmetic-matrix-multiplication-and-comparison-operations) still say it ignores its third argument).

Modular matrix exponentiation would go down a separate path.  Indeed, matrix multiplication normally goes through BLAS, which doesn't have this functionality as far as I know.  I'm not one of the NumPy maintainers, so they'd be the authority, but I suspect you'll want to file a separate feature request for that (referencing this issue if you like).

If you want a workaround, you might be able to do something similar to the code I wrote above, but with `@` instead of `*`.  That said, I haven't looked at it; I'm not sure what guarantees matrix multiplication makes under a modular congruence, so I can't be sure that exponentiation by squaring like this will work.

**Comment by piquan** [comment]
> Never thought I would need this until I did hahaha. Is there an efficient workaround for array exponents?

Well, if you still need this (over five years later), then yes, there is.  You can use essentially the same algorithm, but replace the while condition by using `np.any`, and instead of having the `result=` be in an `if` block, use `np.where`.  The efficiency may go down if your exponents vary by several orders of magnitude or are mostly powers of two (it ends up computing and discarding a lot of intermediates), but in the general case you should be good.

**Comment by sh1boot** [comment]
> If you want a workaround, you might be able to do something similar to the code I wrote above, but with `@` instead of `*`.  That said, I haven't looked at it; I'm not sure what guarantees matrix multiplication makes under a modular congruence, so I can't be sure that exponentiation by squaring like this will work.

I was not aware of the @ operator, I was using the matrix type and its * overload in my workaround, which I now see is deprecated.

Unfortunately, with a deprecated matrix type there's not going to be clean syntax to write pow(mat, n, m).  I don't know if @@ is a thing, but that can't take a mod operand anyway so it's no help.

Modular matrix exponentiation works fine this way, anyway.  I ended up testing it pretty thoroughly when I accidentally demoted an argument to an array and got confusing results and built a big wtf testbench to prove my own sanity.

**Comment by piquan** [comment]
There isn't an `@@` operator in Python; you can see the list (as of 3.13) in https://docs.python.org/3.13/reference/lexical_analysis.html#operators

The likely way you'd create the syntax is with a `matpow` method and probably a corresponding function in the np package, to correspond with the `matmul` method and function.  These were common ways to do matrix multiplications prior to Python 3.5, which is the version that introduced the `@` operator.

Of course, in your case, you'd probably write a `matpow` function in your own program and call that.

**Comment by sh1boot** [comment]
FTR I pressed on with my matrix multiplication implementation but ended up investing time in the awkward space where the modulo was large enough to overflow uint64 after the multiplication while the results would still fit, and had to do some coarse performance hacks which should be applicable (but simpler) to vector multiplication as well:

https://www.tīkōuka.dev/extended-precision-numpy-matpow/

This might help as a starter if anyone's going to do a full implementation.
